libs/wampcc/wamp_session.cc

```diff
@@ -877,47 +872,71 @@ void wamp_session::handle_HELLO(json_array& ja)
 
-  /* For now the only supported method is 'wampcra'. */
-  if (intersect.find(WAMP_WAMPCRA) == intersect.end())
-    throw auth_error(WAMP_ERROR_AUTHENTICATION_FAILED,
-                     "wampcra not available for both client and server");
+  /* Selecte the first method in the set as the common one */
+  std::string authmethod;
+  for(auto &element : intersect) {
+    authmethod = element;
+    break;
+  }
 
-  /* --- Perform wampcra authentication --- */
+  LOG_INFO(m_log_prefix << "selected '" << authmethod << "', as the common authentication method");
 
-  json_object challenge;
-  challenge["nonce"] = random_ascii_string(30);
-  challenge["authprovider"] = m_auth_proivder.provider_name(realm);
-  challenge["authid"] = authid;
-  challenge["timestamp"] = iso8601_utc_timestamp();
-  challenge["authrole"] = "user";
-  challenge["authmethod"] = WAMP_WAMPCRA;
-  challenge["session"] = std::to_string( unique_id() );
-  std::string challengestr = json_encode( challenge );
+  json_object extra;
 
-  {
-    std::lock_guard<std::mutex> guard(m_realm_lock);
-    if (m_challenge.empty())
-      m_challenge = challengestr;
-    else
-      throw auth_error(WAMP_ERROR_AUTHENTICATION_FAILED,
-                       "challenge already issued");
-  }
+  if(m_auth_proivder.hello) {
+    extra = m_auth_proivder.hello(authid, realm, authmethod, m_auth_proivder.provider_name(realm), unique_id() );
+  } else {
+    if(authmethod == WAMP_WAMPCRA) {
+      /* --- Perform wampcra authentication --- */
+
+      json_object challenge;
+      challenge["nonce"] = random_ascii_string(30);
+      challenge["authprovider"] = m_auth_proivder.provider_name(realm);
+      challenge["authid"] = authid;
+      challenge["timestamp"] = iso8601_utc_timestamp();
+      challenge["authrole"] = authrole;
+      challenge["authmethod"] = WAMP_WAMPCRA;
+      challenge["session"] = std::to_string( unique_id() );
+      std::string challengestr = json_encode( challenge );
 
+      {
+        std::lock_guard<std::mutex> guard(m_realm_lock);
+        if (m_challenge.empty())
+          m_challenge = challengestr;
+        else
+          throw auth_error(WAMP_ERROR_AUTHENTICATION_FAILED,
+                           "challenge already issued");
+      }
 
-  json_object extra;
-  extra["challenge"] = std::move(challengestr);
+      extra["challenge"] = std::move(challengestr);
+
+      // attach salting parameters if available
+      if (m_auth_proivder.cra_salt) {
+        auto salt = m_auth_proivder.cra_salt(realm, authid);
+        extra.insert({"salt",salt.salt});
+        extra.insert({"keylen",salt.keylen});
+        extra.insert({"iterations",salt.iterations});
+      }
 
-  // attach salting parameters if available
-  if (m_auth_proivder.cra_salt) {
-    auto salt = m_auth_proivder.cra_salt(realm, authid);
-    extra.insert({"salt",salt.salt});
-    extra.insert({"keylen",salt.keylen});
-    extra.insert({"iterations",salt.iterations});
+      {
```

Comment:
This scope, which acquires & releases mutex, what is its purpose?

Reply:
That guard is from the [master branch](https://github.com/darrenjs/wampcc/blob/master/libs/wampcc/wamp_session.cc#L897) not sure why it is needed.

Reply:
okay ... I'll remove that later